xks-axum/src/xks_proxy/pkcs11.rs

```diff
@@ -51,11 +56,87 @@ use crate::{settings, SETTINGS};
 
 const P11_CONTEXT_READ_FAILURE: &str = "Failed to acquire pkcs11 context";
 
+// https://stackoverflow.com/questions/69350600/how-to-reconcile-rust-mutex-and-c-caller-supplied-locking-mechanisms
+
+#[repr(C)]
+struct MutexContainer {
+    mutex: Mutex<()>,
+    guard: Cell<Option<MutexGuard<'static, ()>>>,
+}
+
+extern "C" fn create_mutex_container(ppcontainer: CK_VOID_PTR_PTR) -> CK_RV {
+    unsafe {
+        *ppcontainer = Box::into_raw(Box::new(MutexContainer {
+            mutex: Mutex::new(()),
+            guard: Cell::new(None),
+        }))
+        .cast();
+        tracing::trace!("create_mutex_container: *ppcontainer: {:?}", *ppcontainer);
+    }
+    CKR_OK
+}
+
+extern "C" fn destroy_mutex_container(pcontainer: CK_VOID_PTR) -> CK_RV {
+    tracing::trace!("destroy_mutex_container: pmutex_container: {pcontainer:?}");
+    unlock_mutex(pcontainer);
+    unsafe {
+        drop(Box::from_raw(pcontainer.cast::<MutexContainer>()));
+    }
+    CKR_OK
+}
+
+extern "C" fn lock_mutex(pcontainer: CK_VOID_PTR) -> CK_RV {
+    tracing::trace!("lock_mutex: pmutex_container: {pcontainer:?}");
+    let container: &MutexContainer = unsafe { &*pcontainer.cast() };
+    match container.mutex.lock() {
+        Ok(guard) => {
+            if let Some(existing) = container.guard.take() {
+                tracing::warn!("pmutex_container {pcontainer:?} already locked: existing={existing:?}, guard={guard:?}");
+            }
+            container.guard.set(Some(guard));
+            CKR_OK
+        }
+        Err(err) => {
+            tracing::warn!("lock_mutex failure: {err}");
+            CKR_MUTEX_BAD
+        }
+    }
+}
+
+extern "C" fn unlock_mutex(pcontainer: CK_VOID_PTR) -> CK_RV {
+    tracing::trace!("unlock_mutex: pmutex_container: {pcontainer:?}");
+    let container: &MutexContainer = unsafe { &*pcontainer.cast() };
+    if let Ok(guard) = container.mutex.try_lock() {
+        std::mem::drop(guard);
+        CKR_MUTEX_NOT_LOCKED
+    } else {
+        let guard = container.guard.take();
+        std::mem::drop(guard);
+        CKR_OK
+    }
```

Comment:
Note that this does not correctly guard against thread A unlocking a mutex it hasn't locked which is locked by thread B. In that case, `try_lock` returns `Err(WouldBlock)` and the guard (that thread B created) is dropped (in a racy way) by thread A.

Reply:
Very true.  The current logic relies on the underlying HSM to do the right thing.

Thinking out load, how about storing a unique thread id of the thread that acquires the lock so that it can be checked against the thread that does the unlock?

On the other hand, the PKCS#11 spec doesn't say anything about if a mutex can be legitimately unlocked from a thread different from the original locking thread.

Reply:
Accessing the thread ID would require to lock the mutex, unless atomic operations are used to access it. Given that the pkcs11 specification states such a call is undefined, it means it's a bug, and so the entire unlock function could just drop the guard unconditionally (without trying to lock first). Trying to lock protects only for the case where a mutex is unlocked and no other thread has locked it, so it does not seem that useful in a multi-threaded scenario where the mutex could be locked at any time by another thread. The status code to return could be determined from `guard.is_some()`.

The current code does not hurt however -- it might just give a somewhat false sense of protection.

Reply:
Good point.  I am inclined to remove the false protection and keep the code simple.

Reply:
Please take a look at the latest revision.

Reply:
Looks good. Thank you!